### PR TITLE
Reduce header dependencies

### PR DESCRIPTION
**Pull request by malfet**
This stack of 3 commits is aimed at showing how some forward declaration can be used to reduce compilation time.
Two second win is indeed insignificant, but it shows the potential compilation speedups by removing unused dependencies.

### CODE REVIEW

File: aten/src/ATen/native/ReflectionPad.cpp [hunk 1]
```diff
@@ -479,7 +480,7 @@ Tensor& reflection_pad1d_out_cpu(
 }
 
 Tensor reflection_pad1d_cpu(const Tensor& input, IntArrayRef padding) {
-  auto output = at::empty({0}, input.options());
+  auto output = at::native::empty({0}, c10::optional<DimnameList>(), input.options());
```
Comment (ilia-cher) [inline aten/src/ATen/native/ReflectionPad.cpp]: curious why did you need to change argument list here (and other places)

Comment (malfet) [reply aten/src/ATen/native/ReflectionPad.cpp]: Unlike `at::empty()`,  `at::native::empty()` needs three arguments. And aren't functions defined in `NativeFunctions.h` one level of indirection less than their `Functions.h` counterparts?

**Review by ilia-cher — APPROVED** [review]

File: aten/src/ATen/native/ReflectionPad.cpp (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/native/ReflectionPad.cpp]: Although it's OK in the cases of this PR, in *general*, use of the `native` version over the `at` version is not semantics preserving.  You need to reason that it is indeed OK. There are few things that `at` does over `native`, currently it's mostly switching devices and handling variable, both of which are not relevant here.
